homeassistant/components/sonos/binary_sensor.py

```diff
@@ -0,0 +1,87 @@
+"""Entity representing a Sonos power sensor."""
+from __future__ import annotations
+
+import datetime
+import logging
+from typing import Any
+
+from homeassistant.components.binary_sensor import (
+    DEVICE_CLASS_BATTERY_CHARGING,
+    DOMAIN as BINARY_SENSOR_DOMAIN,
+    BinarySensorEntity,
+)
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_connect,
+    async_dispatcher_send,
+)
+
+from .const import DATA_SONOS, SONOS_DISCOVERY_UPDATE, SONOS_ENTITY_CREATED
+from .entity import SonosEntity
+from .speaker import SonosSpeaker
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_BATTERY_POWER_SOURCE = "power_source"
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up Sonos from a config entry."""
+
+    async def _async_create_entity(speaker: SonosSpeaker) -> SonosPowerEntity | None:
+        if speaker.battery_info:
+            return SonosPowerEntity(speaker, hass.data[DATA_SONOS])
+        return None
+
+    async def _async_create_entities(speaker: SonosSpeaker):
+        if entity := await _async_create_entity(speaker):
+            async_add_entities([entity])
+        else:
+            async_dispatcher_send(
+                hass, f"{SONOS_ENTITY_CREATED}-{speaker.soco.uid}", BINARY_SENSOR_DOMAIN
+            )
+
+    config_entry.async_on_unload(
+        async_dispatcher_connect(hass, SONOS_DISCOVERY_UPDATE, _async_create_entities)
+    )
+
+
+class SonosPowerEntity(SonosEntity, BinarySensorEntity):
+    """Representation of a Sonos power entity."""
+
+    async def async_added_to_hass(self) -> None:
+        """Register polling callback when added to hass."""
+        await super().async_added_to_hass()
+        async_dispatcher_send(
+            self.hass, f"{SONOS_ENTITY_CREATED}-{self.soco.uid}", BINARY_SENSOR_DOMAIN
```

Comment:
Do we still need this signal?

Reply:
Yes and no. We do need to wait for all entities of a given speaker to be created before subscribing to event callbacks. However this can be optimized to avoid dispatching into both sensor platforms (and back) if the speaker doesn't have a battery.